upt.1.md
---
title: UPT(1)
author: Cyril Roelandt <upt@framalistes.org>
date: DATEPLACEHOLDER
---

# NAME

upt - Universal Packaging Tool

# SYNOPSIS

upt list-backends

upt list-frontends

upt package [-f FRONTEND] [-b BACKEND] [-o OUTPUT] [-u/\--update] [\--debug] [\--quiet] PACKAGE[@VERSION]

# DESCRIPTION
Create a package for a distribution (such as OpenBSD, Fedora, etc.) from a
language-specific package archive (such as the Python Package Index, the
Comprehensive Perl Archive Network, etc.).

Upt works with two kinds of modules:

- frontends, which parse package metadata from a software archive;

- backends, which create package definitions suitable for inclusion in your
  favorite distribution.

The following subcommands are available:

**list-backends**
: List all installed backends.

**list-frontends**
: List all installed frontends.

**package [options...] \<package>[@version]**
: Package the given package. This usually requires options, described below.
  If no version is specified, the latest one is used.


# OPTIONS
-b *BACKEND*, \--backend *BACKEND*

: Specify backend to use. Must be one of the values returned by "upt
list-backends". At least one backend must be installed. If only one backend is
installed, this is optional; otherwise it is required. See BACKENDS.

-c, \--color

: Show colored logging output. Requires the "colorlog" package, available
through the "color" extra.

\--debug

: Print debug messages. Cannot be used with \--quiet.

-f *FRONTEND*, \--frontend *FRONTEND*

: Specify frontend to use. Must be one of the values returned by "upt
list-frontends". At least one frontend must be installed. If only one frontend
is installed, this is optional; otherwise it is required. See FRONTENDS.

-h, \--help

: Show help and return.

-o, \--output *OUTPUT*

: Specify an output file or directory. If this option is not specified, stdout
will be used (if possible). The exact meaning of this option may vary depending
on the backend: it should do the most natural thing.

-q, \--quiet

: Suppress all logging output. Cannot be used with \--debug.

-r, \--recursive

: Recursively package requirements.

-u, \--update

: Update a package.


# BACKENDS
**upt-fedora**
: Create packages for Fedora

**upt-freebsd**
: Create packages for FreeBSD

**upt-guix**
: Create packages for GNU Guix

**upt-macports**
: Create packages for MacPorts

**upt-nix**
: Create packages for Nix

**upt-openbsd**
: Create packages for OpenBSD

# FRONTENDS
**upt-cpan**
: Gather metadata about packages hosted on cpan.org

**upt-pypi**
: Gather metadata about packages hosted on pypi.org

**upt-rubygems**
: Gather metadata about packages hosted on rubygems.org

# EXAMPLES

**List installed frontends**

: upt list-frontends

**List installed backends**

: upt list-backends

**Package "requests" from PyPI, for use in GNU Guix**

: upt package -f pypi -b guix requests

**Same, omitting "-b" when there is only one installed backend**

: upt package -f pypi requests

**Same, omitting "-f" when there is only one installed frontend**

: upt package requests


**Same, but package a given version**

: upt package requests@2.16.0

**Update an existing package**

: upt package -f pypi -b macports -u requests

**Update an existing package to a specific version**

: upt package -f pypi -b macports -u requests@2.22.0

# BUGS

Bugs can be reported to upt@framalistes.org. A web interface is also available
at https://framagit.org/upt (one may sign in using their
github.com/gitlab.com/bitbucket.org credentials). Feel free to suggest new
backends and frontends.

Users are also welcome in the #upt-packaging channel on Freenode.


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

Based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [Unreleased]
### Changed
- colorlog is no longer a required dependency. It can be installed by using the
  "color" extra.

### Fixed
- Fixed error messages for InvalidPackageNameError and
  InvalidPackageVersionError.

## [0.11.1] - 2021-05-02
### Fixed
- Add missing dependency on setuptools

## [0.11] - 2019-12-09
### Added
- New option: --color, to show colored logs
- Allow users to specify a specific version of the package on the CLI
  using "package@version"

## [0.10.3] - 2019-07-31
### Fixed
- The "wheel" package is no longer required when not building a wheel

## [0.10] - 2019-07-21
### Added
- A man page is now available
- New option: -r/--recursive, to recursively package all requirements
- New checksum available: SHA512

## [0.9] - 2019-06-19
### Added
- licenses: Add new licenses
- upt.licenses: Expose get_license_by_spdx_identifier

## [0.8] - 2019-06-15
### Changed
- cli: --frontend/--backend is optional when there is only one
  frontend/backend available
- ArchiveType: add type "RUBYGEM"

## [0.7] - 2019-04-28
### Added
- upt.Archive: add a "filename" attribute
- upt.Archive: add support for sha256 in base64 notation

### Removed
- upt.Package: remove the download_urls attribute

## [0.6] - 2019-02-08
### Added
- upt.Archive: a representation of archive files, such as source tarballs

### Deprecated
- upt.Package.download_urls has been deprecated. Users should use
  upt.Package.archives instead.

## [0.5]
### Fixed
- Added the requirements.txt file
- Officially support more backends

## [0.4.1] - 2018-08-13
- upt.log: Add missing file
- README.md: Fix logger documentation

## [0.4] - 2018-03-20
### Added
- upt package: Add the '--output' option
- upt.log: New module that provides frontends/backends with a logger
- upt-nix is now an officially supported backend

## [0.3] - 2018-03-13
### Added
- Added 'frontends' and 'backends' extras to make installation easier

## [0.2.2] - 2018-03-04
### Fixed
- upt.licenses.License: add __eq__ so that comparisons do not fail

### Changed
- Use setup.cfg

## [0.2.1] - 2018-03-04
### Added
- upt.licenses: Add the OpenSSL license
- upt.licenses: Add the Perl license
- upt.licenses: Add the Ruby license
- upt.licenses: Add the GFDL (1.1, 1.2, 1.3) license
- upt.licenses: Add the Zlib license

## [0.2] - 2018-02-21
### Added
- upt.Package: Add a 'licenses' field

### Fixed
- Fixed installation instructions in README.md
- Add missing files to the tarball (LICENSE, CHANGELOG, README.md, tox.ini)

## [0.1.1] - 2018-02-17
### Fixed
- Fixed flake8 target in tox.ini

## [0.1] - 2018-02-17
### Added
- Initial CLI tool (upt)
- Initial library
- Support for user-defined frontends
- Support for user-defined backends


upt/upt.py
# Copyright 2018      Cyril Roelandt
#
# Licensed under the 3-clause BSD license. See the LICENSE file.
import argparse
from enum import Enum
import logging
import os
import urllib.request
import sys
import tempfile

from packaging.specifiers import SpecifierSet
import pkg_resources

import upt.checksum
import upt.exceptions
import upt.log


class Backend(object):
    def current_version(self, frontend, pkgname, output=None):
        """Return the currently packaged version of a package.

        This method must return the currently packaged version of PKGNAME,
        packaged using the given FRONTEND.
        TODO: output
        """
        return input(f'Current version of {pkgname} ({frontend.name})?')

    def update_package(self, pdiff, output=None):
        raise NotImplementedError()

    def needs_requirement_interactive(self, req, phase):
        while True:
            ret = input(f'Should we package {req} (phase {phase})? [y/n] ')
            if ret.lower() in ('y', 'yes'):
                return True
            elif ret.lower() in ('n', 'no'):
                return False
            else:
                continue  # pragma: nocover

    def needs_requirement(self, req, phase):
        logger = logging.getLogger('upt')
        if not hasattr(self, 'package_versions'):
            return self.needs_requirement_interactive(req, phase)

        versions = self.package_versions(req.name)
        if not versions:
            logger.info(f'Dependency {req}: currently not packaged. '
                        f'Packaging it.')
            return True

        if not req.specifier:
            logger.info(f'Dependency {req}: currently packaged, no specific '
                        f'version is required. Not packaging it.')
            return False

        s = SpecifierSet(req.specifier)
        compatible_versions = list(s.filter(versions))
        if compatible_versions:
            logger.info(f'Dependency {req}: at least one compatible version '
                        f'found in {compatible_versions}. Not packaging it.')
            return False
        else:
            logger.info(f'Dependency {req}: no compatible version found '
                        f'in {versions}. Packaging it.')
            return True


class Frontend(object):
    """Base class for all upt Frontends."""
    pass


class PackageUpToDateException(Exception):
    def __init__(self, pkgname, version):
        self.pkgname = pkgname
        self.version = version

    def __str__(self):
        return f'{self.pkgname} is already up-to-date ({self.version})'


class ArchiveType(Enum):
    SOURCE_TARGZ = 1
    RUBYGEM = 2


class ArchiveUnavailable(Exception):
    def __str__(self):
        return 'No such archive could be found'


class Archive(object):
    '''An archive file.

    This can be a source tarball, a Python wheel, a Ruby gem, a binary, etc.
    '''
    def __init__(self, url, archive_type=ArchiveType.SOURCE_TARGZ, size=0,
                 md5=None, sha256=None, rmd160=None, sha256_base64=None,
                 sha512=None):
        self.url = url
        self.archive_type = archive_type
        self._size = size
        self._filepath = None  # Absolute path on the local filesystem
        self._filename = None  # Filename on the local filesystem
        self._hashes = {}
        if md5 is not None:
            self.md5 = md5
        if sha256 is not None:
            self.sha256 = sha256
        if rmd160 is not None:
            self.rmd160 = rmd160
        if sha256_base64 is not None:
            self.sha256_base64 = sha256_base64
        if sha512 is not None:
            self.sha512 = sha512

    @property
    def filepath(self):
        if self._filepath is None:
            self._filepath = os.path.join(tempfile.gettempdir(), self.filename)
            urllib.request.urlretrieve(self.url, filename=self._filepath)
        return self._filepath

    @property
    def filename(self):
        if self._filename is None:
            self._filename = os.path.basename(self.url)
        return self._filename

    @property
    def size(self):
        if self._size == 0:
            self._size = os.stat(self.filepath).st_size
        return self._size

    def _checksum(self, hash_name):
        try:
            return self._hashes[hash_name]
        except KeyError:
            value = upt.checksum.compute_checksum(self.filepath, hash_name)
            self._hashes[hash_name] = value
            return value

    @property
    def md5(self):
        return self._checksum('md5')

    @md5.setter
    def md5(self, value):
        self._hashes['md5'] = value

    @property
    def rmd160(self):
        return self._checksum('rmd160')

    @rmd160.setter
    def rmd160(self, value):
        self._hashes['rmd160'] = value

    @property
    def sha256(self):
        return self._checksum('sha256')

    @sha256.setter
    def sha256(self, value):
        self._hashes['sha256'] = value

    @property
    def sha256_base64(self):
        return self._checksum('sha256_base64')

    @sha256_base64.setter
    def sha256_base64(self, value):
        self._hashes['sha256_base64'] = value

    @property
    def sha512(self):
        return self._checksum('sha512')

    @sha512.setter
    def sha512(self, value):
        self._hashes['sha512'] = value


class PackageRequirement(object):
    """Package Requirement.

    Two main attributes are used:
    - name: a string, representing the name of the package.
    - specifier: a string, representing a version specifier as defined in
                 https://www.python.org/dev/peps/pep-0440/#version-specifiers .
                 The format of the specifier is subject to changes.
    """
    def __init__(self, name, specifier=None):
        self.name = name
        self.specifier = specifier

    def __str__(self):
        if self.specifier is not None:
            return f'{self.name} ({self.specifier})'
        else:
            return f'{self.name}'

    def __eq__(self, other):
        return (self.name == other.name and
                self.specifier == other.specifier)


class Package(object):
    """Base class for all packages.

    Constructor arguments:
    - name: required, the name of the package
    - version: required, the version of the package, as a str

    Optional keyword arguments:
    - homepage: a string, representing the official homepage of the package
    - summary: a short summary of the package's purpose
    - description: a long description of the package's purpose
    - requirement: a dict with 3 keys:
        - 'config': the config dependencies of the package;
        - 'build': the build dependencies of the package;
        - 'run': the runtime dependencies of the package;
        - 'test': the test dependencies of the package.
      The values associated with these keys are always lists of
      PackageRequirement objects.
    - licenses: a list of upt.licenses.License objects
    """
    def __init__(self, name, version, **kwargs):
        self.name = name
        self.version = version
        self.homepage = kwargs.get('homepage', '')
        self.summary = kwargs.get('summary', '')
        self.description = kwargs.get('description', '')
        # Software requirements, as instances of PackageRequirement: {
        #     'build': [pkg1, ...],
        #     'run': [pkg2, ...],
        #     'test': [pkg3, ...]
        # }
        self.requirements = kwargs.get('requirements', {})
        self.licenses = kwargs.get('licenses', [])
        self.archives = kwargs.get('archives', [])

    def __str__(self):
        return f'{self.name}@{self.version}'

    def get_archive(self, archive_type=ArchiveType.SOURCE_TARGZ):
        for archive in self.archives:
            if archive.archive_type == archive_type:
                return archive
        else:
            raise ArchiveUnavailable()

    def _clean(self):
        for archive in self.archives:
            if archive._filepath is not None:
                os.remove(archive._filepath)


class PackageDiff(object):
    def __init__(self, old, new):
        self.old = old
        self.new = new

    @property
    def old_version(self):
        return self.old.version

    @property
    def new_version(self):
        return self.new.version

    def new_requirements(self, phase):
        old_names = [req.name for req in self.old.requirements.get(phase, [])]
        new_names = [req.name for req in self.new.requirements.get(phase, [])]
        added_names = set(new_names) - set(old_names)
        return [
            req for req in self.new.requirements.get(phase, [])
            if req.name in added_names
        ]

    def updated_requirements(self, phase):
        old_names = [req.name for req in self.old.requirements.get(phase, [])]
        new_names = [req.name for req in self.new.requirements.get(phase, [])]
        updated_names = set(old_names) & set(new_names)
        return [
            req for req in self.new.requirements.get(phase, [])
            if req.name in updated_names
        ]

    def deleted_requirements(self, phase):
        old_names = [req.name for req in self.old.requirements.get(phase, [])]
        new_names = [req.name for req in self.new.requirements.get(phase, [])]
        deleted_names = set(old_names) - set(new_names)
        return [
            req for req in self.old.requirements.get(phase, [])
            if req.name in deleted_names
        ]


def create_parser(frontends, backends):
    parser = argparse.ArgumentParser(prog='upt')

    subparsers = parser.add_subparsers(title='Commands', dest='cmd')

    # List backends
    subparsers.add_parser('list-backends',
                          help='List installed backends')

    # List frontends
    subparsers.add_parser('list-frontends',
                          help='List installed frontends')

    # Actually package software
    parser_package = subparsers.add_parser('package',
                                           help='Package a piece of software')
    required_args = parser_package.add_argument_group('Required arguments')
    required_args.add_argument('-f', '--frontend',
                               required=len(frontends) > 1,
                               choices=frontends,
                               default=frontends[0] if frontends else None,
                               help='Frontend to use')
    required_args.add_argument('-b', '--backend',
                               required=len(backends) > 1,
                               choices=backends,
                               default=backends[0] if backends else None,
                               help='Backend to use')
    logger_group = parser_package.add_mutually_exclusive_group()
    logger_group.add_argument('--debug', action='store_const',
                              const=logging.DEBUG, dest='log_level',
                              help='Print debug messages.')
    parser_package.add_argument('-o', '--output',
                                help='Output file/directory. Defaults to '
                                     'stdout/current directory. This may be '
                                     'ignored by the backend.')
    logger_group.add_argument('-q', '--quiet', action='store_const',
                              const=logging.CRITICAL+1, dest='log_level',
                              help='Suppress all logging output')
    parser_package.add_argument('-r', '--recursive', action='store_true',
                                help='Recursively package requirements'),
    parser_package.add_argument('-c', '--color', action='store_true',
                                help='Show colored logging output'),
    parser_package.add_argument('-u', '--update', action='store_true',
                                help='Update package'),
    parser_package.add_argument('package', help='Name of the package. '
                                                'Use <package>@<version> to '
                                                'package a specific version.')

    return parser


def _get_installed_plugins(entrypoint):
    plugins = {}
    for ep in pkg_resources.iter_entry_points(entrypoint):
        plugin_cls = ep.load()
        plugins[plugin_cls.name] = plugin_cls

    return plugins


def _get_installed_frontends():
    return _get_installed_plugins('upt.frontends')


def _get_installed_backends():
    return _get_installed_plugins('upt.backends')


def package(pkg_name, version, frontend, backend, output, recursive,
            packaged=[]):
    logger = logging.getLogger('upt')

    try:
        upt_pkg = None
        upt.log.logger_set_formatter(logger, 'Frontend')
        upt_pkg = frontend.parse(pkg_name, version)
        upt_pkg.frontend = frontend.name
        upt.log.logger_set_formatter(logger, 'Backend')
        backend.create_package(upt_pkg, output=output)
        packaged.append(pkg_name)
        if recursive:
            for phase, requirements in upt_pkg.requirements.items():
                for requirement in requirements:
                    # Let's not repackage what we have already packaged.
                    if requirement.name in packaged:
                        continue

                    if backend.needs_requirement(requirement, phase):
                        package(requirement.name, None, frontend, backend,
                                output, recursive, packaged)
    except upt.exceptions.UnhandledFrontendError as e:
        print(e, file=sys.stderr)
        sys.exit(1)
    except (upt.exceptions.InvalidPackageNameError,
            upt.exceptions.InvalidPackageVersionError) as e:
        print(e, file=sys.stderr)
        sys.exit(1)
    finally:
        if upt_pkg is not None:
            # We always want to clean up after ourselves.
            upt_pkg._clean()


def _extract_name_version_from_package(package):
    try:
        name, version = package.split('@')
    except ValueError:
        name, version = package, None
    return name, version or None


def update(pkg_name, version, frontend, backend, output):
    logger = logging.getLogger('upt')

    upt.log.logger_set_formatter(logger, 'Backend')
    old_version = backend.current_version(frontend, pkg_name, output)

    # Parse both the old and the new version of the package in order to create
    # a "diff".
    upt.log.logger_set_formatter(logger, 'Frontend')
    old_pkg = frontend.parse(pkg_name, old_version)
    old_pkg.frontend = frontend.name
    new_pkg = frontend.parse(pkg_name, version)
    new_pkg.frontend = frontend.name
    diff = PackageDiff(old_pkg, new_pkg)
    if diff.new_version == old_version:
        raise PackageUpToDateException(pkg_name, old_version)

    # Let the backend do its magic.
    upt.log.logger_set_formatter(logger, 'Backend')
    logger.info(f'Updating {pkg_name} from {old_version} to {new_pkg.version}')
    try:
        backend.update_package(diff, output=output)
    except NotImplementedError:
        logger.error(f'The "{backend.name}" backend  does not implement the '
                     'update feature yet.')
        sys.exit(1)


def main():
    frontends = _get_installed_frontends()
    backends = _get_installed_backends()
    parser = create_parser(list(frontends.keys()), list(backends.keys()))

    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)

    # We need to handle this here, because if we let the parser run in an
    # environment without frontends/backends, it will return a useless error
    # message, such as "choose a frontend in {}".
    if sys.argv[1] == 'package':
        if not frontends:
            sys.exit('You need to install at least one frontend, for instance '
                     'upt-cpan, upt-pypi or upt-rubygems')

        if not backends:
            sys.exit('You need to install at least one backend, for instance '
                     'upt-fedora, upt-freebsd, upt-guix, upt-nix or '
                     'upt-openbsd')

    args = parser.parse_args()

    if args.cmd == 'list-backends':
        for backend in backends.keys():
            print(backend)
        sys.exit(0)

    if args.cmd == 'list-frontends':
        for frontend in frontends.keys():
            print(frontend)
        sys.exit(0)

    if args.cmd == 'package':
        # There will not be a KeyError here, since argparse will catch "wrong"
        # values for us.
        frontend = frontends[args.frontend]()
        backend = backends[args.backend]()
        name, version = _extract_name_version_from_package(args.package)
        upt.log.create_logger(args.log_level, args.color)
        if args.update:
            try:
                update(name, version, frontend, backend, args.output)
            except PackageUpToDateException as e:
                logger = logging.getLogger('upt')
                logger.info(e)
            except (upt.exceptions.InvalidPackageNameError,
                    upt.exceptions.InvalidPackageVersionError) as e:
                logger = logging.getLogger('upt')
                upt.log.logger_set_formatter(logger, 'Frontend')
                logger.error(e)
                sys.exit(1)
        else:
            package(name, version, frontend, backend, args.output,
                    args.recursive)
